src/MinimalDelegation.sol

```diff
@@ -140,24 +139,19 @@ contract MinimalDelegation is
             }
         }
 
-        (bytes32 keyHash, bytes memory signature, bytes memory hookData) =
-            abi.decode(wrappedSignature, (bytes32, bytes, bytes));
+        (bytes32 keyHash, bytes calldata signature, bytes memory hookData) =
+            wrappedSignature.decodeSignatureWithKeyHashAndHookData();
 
         Key memory key = getKey(keyHash);
-
-        /// There are 3 ways to validate a signature through ERC-1271:
-        /// 1. The caller is allowlisted, so we can validate the signature directly against the data.
-        /// 2. The caller is address(0), meaning it is an offchain call, so we can validate the signature as if it is a PersonalSign.
-        /// 3. If none of the above is true, the signature must be validated as a TypedDataSign struct according to ERC-7739.
-        bool isValid;
-        if (erc1271CallerIsSafe[msg.sender]) {
-            isValid = key.verify(digest, signature);
-        } else if (msg.sender == address(0)) {
-            // We only support PersonalSign for offchain calls
-            isValid = _isValidNestedPersonalSig(key, digest, domainSeparator(), signature);
-        } else {
-            isValid = _isValidTypedDataSig(key, digest, domainBytes(), signature);
-        }
+        /// Signature deduction flow as specified by ERC-7739
+        // If the signature contains enough data for a TypedDataSign, try the TypedDataSign flow
+        bool isValid = _isValidTypedDataSig(key, digest, domainBytes(), signature)
+        // If the signature is not valid as a TypedDataSign, try the NestedPersonalSign flow
+        || (
+            _isValidNestedPersonalSig(key, digest, domainSeparator(), signature)
+            // Finally, if the ERC1271 caller is considered safe, try the raw verification flow
+            || (erc1271CallerIsSafe[msg.sender] && key.verify(digest, signature))
```

Comment:
i wonder if we could actually remove this logic if we introduce raw ecrecover for the eoa itselft

Reply:
Oh sorry I meant to specifically call out the erc1271CallerIsSafe logic

i kind of think we should remove that..it's super jank